### PR TITLE
Define a proof theory for the equational logic of magmas.

### LINKED ISSUE #104: Build the basic meta-theory for the (syntactic) equational theory of magmas
~~No need to bring proofs into this (yet)~~, but this should enable proving compactness, and the meta-theorems alluded to [here](https://leanprover.zulipchat.com/#narrow/stream/458659-Equational/topic/Syntax/near/473455037)

### PR DESCRIPTION
**Pull request by codyroux**
For now just soundness.

Closes #104

### CODE REVIEW

**Comment by teorth** [comment]
Feel free to update the intro.tex chapter of the blueprint to align it with your new contributions, for instance adding new theorems to the blueprint and linking them to your code using the \lean{} keywords (and \uses and \leanok as appropriate).  I'll add some text to CONTRIBUTING.md to explain these macros now, since I realize that perhaps not all contributors are familiar with blueprint.

**Comment by codyroux** [comment]
Ok proved completeness as well, sadly uses choice. Should be avoidable, but I'm gonna take the win for now.

**Comment by codyroux** [comment]
I'm going to get started on the blueprint modifications, if I can remember how to write LaTeX.

File: equational_theories/Completeness.lean [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+-- We keep this in type, because we're going to want to gather
+-- the (finite!) set of required axioms later.
+inductive derive {α} : Ctx α → MagmaLaw α → Type :=
+| Ax Γ E (h : E ∈ Γ) : derive Γ E
+| Ref Γ t : derive Γ (t ≃ t)
+| Sym Γ t u : derive Γ (t ≃ u) → derive Γ (u ≃ t)
+| Trans Γ t u v : derive Γ (t ≃ u) → derive Γ (u ≃ v) → derive Γ (t ≃ v)
+-- This is not as polymorphic as it could be, shouldn't be an issue at the moment
+| Subst Γ t u σ : derive Γ (t ≃ u) → derive Γ (t ⬝ σ ≃ u ⬝ σ)
+| Cong Γ t₁ t₂ u₁ u₂ : derive Γ (t₁ ≃ t₂) → derive Γ (u₁ ≃ u₂) → derive Γ (t₁ ⋆ u₁ ≃ t₂ ⋆ u₂)
```
[the remark this answers is not in the document]
Comment (Shreyas4991) [reply equational_theories/Completeness.lean:35]: Basically you get a locally contained version of the notation for `derive` which you can use to make your inductive type look closer to the inference rule notation on paper

[the remark this answers is not in the document]
Comment (codyroux) [reply equational_theories/Completeness.lean:35]: Neat! I'll do this.

[the remark this answers is not in the document]
Comment (Shreyas4991) [reply equational_theories/Completeness.lean:35]: I just tested it. It works. The `section` is to contain the `set_option hygiene false` setting and the local notation.

The other place where you have defined this notation needs to be intact. The `local` notation's scope is restricted to that `section DeriveDef`

File: equational_theories/Completeness.lean [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+
+-- We keep this in type, because we're going to want to gather
+-- the (finite!) set of required axioms later.
+inductive derive {α} : Ctx α → MagmaLaw α → Type :=
+| Ax Γ E (h : E ∈ Γ) : Γ ⊢ E
+| Ref Γ t : Γ ⊢ (t ≃ t)
+| Sym Γ t u : Γ ⊢ (t ≃ u) → Γ ⊢ (u ≃ t)
+| Trans Γ t u v : Γ ⊢ (t ≃ u) → Γ ⊢ (u ≃ v) → Γ ⊢ (t ≃ v)
+-- This is not as polymorphic as it could be, shouldn't be an issue at the moment
+| Subst Γ t u σ : Γ ⊢ (t ≃ u) → Γ ⊢ (t ⬝ σ ≃ u ⬝ σ)
```
Comment (Command-Master) [inline equational_theories/Completeness.lean:48]: Why do you need this? Isn't it just applying `Cong` with one of the sides being `Ref`?

**Comment by codyroux** [comment]
All right this is a mess, and I have compactness in another branch, I'll just re-create a PR.

**Pull request closed without merging**